### PR TITLE
Add hashFinalSaplingRoot to getblocktemplate

### LINKED ISSUE #3248: Add hashFinalSaplingRoot to getBlockTemplate

### PR DESCRIPTION
**Pull request by Eirik0**
Closes #3248

### CODE REVIEW

File: qa/rpc-tests/getblocktemplate.py [hunk 1]
```diff
@@ -49,11 +49,15 @@ def run_test(self):
 
         # Test 5: General checks
         tmpl = node.getblocktemplate()
-        assert(len(tmpl['noncerange']) == 16)
+        assert_equal(16, len(tmpl['noncerange']))
```
Comment (gtank) [inline qa/rpc-tests/getblocktemplate.py]: What's the difference between these two asserts?

Comment (Eirik0) [reply qa/rpc-tests/getblocktemplate.py]: assert_equals is essentially just a wrapper for assert. I think it reads a bit better when we are not just checking a boolean value.

File: qa/rpc-tests/getblocktemplate.py [hunk 2]
```diff
@@ -49,11 +49,15 @@ def run_test(self):
...
 
         # Test 6: coinbasetxn checks
         assert('foundersreward' in tmpl['coinbasetxn'])
         assert(tmpl['coinbasetxn']['required'])
 
+        # Test 7: hashFinalSaplingRoot checks
+        assert('finalsaplingroothash' in tmpl)
+        assert_equal(64, len(tmpl['finalsaplingroothash']))
```
Comment (str4d) [inline qa/rpc-tests/getblocktemplate.py]: This should be the empty Sapling root, so we should be able to compare them directly.

Comment (Eirik0) [reply qa/rpc-tests/getblocktemplate.py]: What do you mean by the empty Sapling root? If I print out ```tmpl['finalsaplingroothash']``` I get some seemingly random hash value (3e49b5f954aa9d3545bc6c37744661eea48d7c34e3000d82b7f0010c30f4c2fb).

Comment (ebfull) [reply qa/rpc-tests/getblocktemplate.py]:
```
    uint256 expected = uint256S("3e49b5f954aa9d3545bc6c37744661eea48d7c34e3000d82b7f0010c30f4c2fb");

    ASSERT_TRUE(ZCSaplingIncrementalMerkleTree::empty_root() == expected);
```

I guess that's good!

**Comment by Eirik0** [comment]
Changed the assert to compare the final sapling root to '3e49b5f954aa9d3545bc6c37744661eea48d7c34e3000d82b7f0010c30f4c2fb', and force pushed that as the commit.

**Comment by ebfull** [comment]
We still need to discuss endianness. What will pool operators need/expect from this output?

**Comment by Eirik0** [comment]
@ebfull We are doing finalsaplingroothash the same way as previousblockhash:
```
result.push_back(Pair("previousblockhash", pblock->hashPrevBlock.GetHex()));
```

**Comment by str4d** [comment]
@zkbot r+

**Pull request merged**